src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1703,17 +1703,14 @@ private static bool IsPropertyAssignedThroughBackingField(BoundExpression receiv
                 propertySymbol = propertySymbol.OriginalDefinition;
             }
 
-            // PROTOTYPE(semi-auto-props): TODO: Support assigning semi auto prop from constructors.
-
             return propertySymbol is SourcePropertySymbolBase sourceProperty &&
-                    // PROTOTYPE(semi-auto-props): Consider `public int P { get => field; set; }`
-                    sourceProperty.GetMethod is SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: true } &&
+                    IsConstructorOrField(fromMember, isStatic: sourceProperty.IsStatic) &&
+                    (sourceProperty.GetMethod is SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: true } || sourceProperty.FieldKeywordBackingField is not null) &&
```

Comment:
This doesn't seem correct. Whether or not the backing field is assigned directly is not dependent on whether `FieldKeywordBackingField` is generated, only on whether it is a get only or `set;` accessor.

Reply:
@333fred The method returns true if the given property is assigned through backing field *directly* or *indirectly*. Meaning that if the `set` accessor is known to assign the backing field, we'll return true. There is a check below for that:

> sourceProperty.SetMethod is null or SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: true }

Anything I'm missing?

Reply:
> Meaning that if the set accessor is known to assign the backing field, we'll return true.

We cannot know this unless the setter either does not exist, or is auto-implemented. If it is neither of these, then assignment in the constructor must call the setter itself.

Reply:
> the setter either does not exist, or is auto-implemented

@333fred Yes, and that's what the code is intended to do:

> sourceProperty.SetMethod is null or SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: true }

Here I'm checking for non-existing setter, or a semicolon-only setter.

Not sure if I'm missing something.

Reply:
Unless I'm missing something, you're checking if it's non existent, semi-colon only, _or_ (the new code) uses the `field` keyword.

Reply:
@333fred The check is:

```csharp
some conditions &&
(sourceProperty.GetMethod is SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: true } || sourceProperty.FieldKeywordBackingField is not null) &&
sourceProperty.SetMethod is null or SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: true } &&
some more conditions
```

There are two checks that are `&&`ed together:

1. Getter is semicolon or there is field keyword backing field.
2. Setter is null or semicolon only.

Both 1 and 2 should must be true for the method to return true.

Reply:
Probably mixing `&&` with `||` in the same expression with parentheses is confusing? Or there is something else I'm missing?

Reply:
It seems like we just need to check: is there a field backing keyword, and is the setter missing or auto-implemented.

Reply:
> It seems like we just need to check: is there a field backing keyword, and is the setter missing or auto-implemented.

Will this maintain the behavior for ```int P {get; set;}```?

Reply:
Anything I need to do here?

Reply:
CC @333fred to follow up on this thread.